xcdat/dataset.py

```diff
@@ -161,22 +185,219 @@ def open_mfdataset(
+    """
+    if decode_times:
+        cf_compliant_time: Optional[bool] = has_cf_compliant_time(paths)
+        # XCDAT handles decoding decoding time values with non-CF units using
+        # the preprocess kwarg.
+        if cf_compliant_time is False:
+            decode_times = False
+            preprocess = partial(_preprocess_non_cf_dataset, callable=preprocess)
+
+    ds = xr.open_mfdataset(
+        paths,
+        decode_times=decode_times,
+        data_vars=data_vars,
+        preprocess=preprocess,
+        **kwargs,
+    )
+    ds = infer_or_keep_var(ds, data_var)
+    ds = ds.bounds.add_missing_bounds()
+    return ds
 
-    Keep multiple variables in the Dataset:
 
-    >>> from xcdat.dataset import open_dataset
-    >>> ds = open_mfdataset(["file_path1", "file_path2"], data_var=["ts", "tas"])
+def has_cf_compliant_time(
+    path: Union[
+        str,
+        pathlib.Path,
+        List[str],
+        List[pathlib.Path],
+        List[List[str]],
+        List[List[pathlib.Path]],
+    ]
+) -> Optional[bool]:
+    """Determine if a dataset has time coordinates with CF compliant units.
+
+    This function opens a single dataset either from a single string path, glob
+    string path, a list, or a list of lists. If the dataset does not contain a
+    time dimension, None is returned. Otherwise, the units attribute is
+    extracted from the time coordinates to determine whether it is CF or
+    non-CF compliant.
+
+    Parameters
+    ----------
+    path : Union[str, pathlib.Path, List[str], List[pathlib.Path], \
+         List[List[str]], List[List[pathlib.Path]]]
+        Either a file (``"file.nc"``), a string glob in the form
+        ``"path/to/my/files/*.nc"``, or an explicit list of files to open.
+        Paths can be given as strings or as pathlib Paths. If concatenation
+        along more than one dimension is desired, then ``paths`` must be a
+        nested list-of-lists (see ``combine_nested`` for details). (A string
+        glob will be expanded to a 1-dimensional list.)
+
+    Returns
+    -------
+    Optional[bool]
+        None if time dimension does not exist, True if CF compliant, or False if
+        non-CF compliant.
+
+    Notes
+    -----
+    This function only checks one file for multi-file datasets to optimize
+    performance because it is slower to combine all files then check for CF
+    compliance.
     """
-    ds = xr.open_mfdataset(paths, decode_times=False, data_vars=data_vars, **kwargs)
-    ds = infer_or_keep_var(ds, data_var)
+    first_file: Optional[Union[pathlib.Path, str]] = None
+    if isinstance(path, str) and "*" in path:
+        first_file = glob(path)[0]
+    elif isinstance(path, str) or isinstance(path, pathlib.Path):
+        first_file = path
+    elif isinstance(path, list):
+        if any(isinstance(sublist, list) for sublist in path):
+            first_file = path[0][0]  # type: ignore
+        else:
+            first_file = path[0]  # type: ignore
```

Comment:
I updated this to support `pathlib.Path` and nested list of strings/`pathlib.Paths`.